I get an error, when I disconnect the learner:
```

---------------------------- IndexError Exception -----------------------------
Traceback (most recent call last):
  File "/Users/vesna/orange-canvas-core/orangecanvas/scheme/signalmanager.py", line 1024, in __process_next
    if self.__process_next_helper(use_max_active=True):
  File "/Users/vesna/orange-canvas-core/orangecanvas/scheme/signalmanager.py", line 1062, in __process_next_helper
    self.process_node(selected_node)
  File "/Users/vesna/orange-canvas-core/orangecanvas/scheme/signalmanager.py", line 690, in process_node
    self.send_to_node(node, signals_in)
  File "/Users/vesna/orange-widget-base/orangewidget/workflow/widgetsscheme.py", line 824, in send_to_node
    self.process_signals_for_widget(node, widget, signals)
  File "/Users/vesna/orange-widget-base/orangewidget/workflow/widgetsscheme.py", line 838, in process_signals_for_widget
    process_signals_for_widget(widget, signals, workflow)
  File "/Users/vesna/miniconda3/lib/python3.8/functools.py", line 875, in wrapper
    return dispatch(args[0].__class__)(*args, **kw)
  File "/Users/vesna/orange-widget-base/orangewidget/workflow/widgetsscheme.py", line 939, in process_signals_for_widget
    process_signal_input(input_meta, widget, signal, workflow)
  File "/Users/vesna/miniconda3/lib/python3.8/functools.py", line 875, in wrapper
    return dispatch(args[0].__class__)(*args, **kw)
  File "/Users/vesna/orange-widget-base/orangewidget/workflow/widgetsscheme.py", line 902, in process_signal_input_default
    notify_input_helper(
  File "/Users/vesna/miniconda3/lib/python3.8/functools.py", line 875, in wrapper
    return dispatch(args[0].__class__)(*args, **kw)
  File "/Users/vesna/orange-widget-base/orangewidget/utils/signals.py", line 756, in set_multi_input_helper
    handler(*args)
  File "/Users/vesna/orange-widget-base/orangewidget/utils/signals.py", line 298, in summarize_wrapper
    id_ = ids.pop(index)
IndexError: pop index out of range
```

To reproduce, consider the workflow in biolab/orange3#5611 and remove the LinearRegression widget.

--- a/orangewidget/utils/signals.py
+++ b/orangewidget/utils/signals.py
@@ -657,7 +657,7 @@ def set_multi_input_helper(
     inout.closing_sentinel as the obj.
     """
     inputs_ = get_widget_inputs(widget)
-    inputs = list(inputs_.setdefault(input.name, ()))
+    inputs = inputs_.setdefault(input.name, ())
     filter_none = input.filter_none
 
     signal_old = None
@@ -672,15 +672,16 @@ def set_multi_input_helper(
         index = key_to_pos.get(key)
         assert index is not None
 
+    inputs_updated = list(inputs)
     if new:
-        inputs.insert(index, (key, obj))
+        inputs_updated.insert(index, (key, obj))
     elif remove:
-        signal_old = inputs.pop(index)
+        signal_old = inputs_updated.pop(index)
     else:
-        signal_old = inputs[index]
-        inputs[index] = (key, obj)
+        signal_old = inputs_updated[index]
+        inputs_updated[index] = (key, obj)
 
-    inputs_[input.name] = tuple(inputs)
+    inputs_[input.name] = tuple(inputs_updated)
 
     if filter_none:
         def filter_f(obj):
@@ -716,6 +717,8 @@ def set_multi_input_helper(
             if filter_f(signal_old[1]):
                 # was already notified as removed, only remove from inputs (done above)
                 return
+            else:
+                index = local_index(key, inputs, filter_f)
         elif update and filtered:
             if filter_f(signal_old[1]):
                 # did not change; remains filtered

--- a/orangewidget/workflow/tests/test_widgetsscheme.py
+++ b/orangewidget/workflow/tests/test_widgetsscheme.py
@@ -528,6 +528,25 @@ class TestSignalManager(GuiTest):
         check_inputs([2])
         model.remove_link(l1)
         check_inputs([2])
+        reset_events()
+
+        model.remove_link(l2)
+        check_inputs([])
+        check_events([("remove", 0)])
+        reset_events()
+
+        w1.Outputs.out.send(None)
+        w2.Outputs.out.send(1)
+        model.insert_link(0, l1)
+        model.insert_link(1, l2)
+        check_inputs([1])
+        check_events([("insert", 0, 1)])
+        reset_events()
+        # ensure proper index on input removal when preceding inputs are
+        # filtered
+        model.remove_link(l2)
+        check_inputs([])
+        check_events([("remove", 0)])
 
     def test_old_style_input(self):
         model, widgets = create_workflow()

